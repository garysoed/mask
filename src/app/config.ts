import { VineImpl } from 'grapevine/export/main';
import { CustomElementCtrl } from 'persona/export/main';

export interface Config {
  /** TODO: Delete ctrl */
  ctor?: typeof CustomElementCtrl;
  dependencies?: Config[];
  tag: string;
  configure?(vine: VineImpl): void;
}
