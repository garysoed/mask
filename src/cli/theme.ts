import * as fs from 'fs';

import commandLineArgs from 'command-line-args';
import {Color, fromCssColor} from 'gs-tools/export/color';
import {EMPTY, Observable, Subject} from 'rxjs';
import {catchError, takeUntil} from 'rxjs/operators';
import {Logger, ON_LOG_$} from 'santa';
import {CliDestination} from 'santa/export/cli';

import {Palette, PALETTE} from '../theme/palette';
import {Theme} from '../theme/theme';


const LOGGER = new Logger('@hive/main');

const OPTIONS = [
  {
    name: 'base',
    type: String,
  },
  {
    name: 'acc',
    type: String,
  },
  {
    name: 'out',
    type: String,
  },
];

const options = commandLineArgs(OPTIONS, {stopAtFirstUnknown: true});

/**
 * Completes whenever running is done.
 */
function run(): Observable<unknown> {
  const base = findColor(options.base);
  const acc = findColor(options.acc);
  const outStr = options.out;
  const theme = new Theme(base, acc);

  return new Observable(subscriber => {
    fs.writeFile(outStr, theme.generateCss(), {encoding: 'utf8'}, err => {
      if (err) {
        subscriber.error(err);
      } else {
        subscriber.next();
        subscriber.complete();
      }
    });
  });
}

function findColor(colorStr: string): Color {
  const existingColor: Color|undefined = PALETTE[colorStr.toUpperCase() as keyof Palette];
  if (existingColor) {
    return existingColor;
  }

  const cssColor = fromCssColor(colorStr);
  if (!cssColor) {
    throw new Error(`Unsupported color: ${colorStr}`);
  }

  return cssColor;
}

const onDone$ = new Subject();
const consoleLog = new CliDestination(
    entry => {
      if (entry.key === 'authurl') {
        return {showKey: false, enableFormat: false};
      }
      return {showKey: false};
    },
);
ON_LOG_$
    .pipe(takeUntil(onDone$))
    .subscribe(entry => {
      consoleLog.log(entry);
    });

run()
    .pipe(
        catchError(e => {
          LOGGER.error(e);
          return EMPTY;
        }),
    )
    .subscribe({complete: () => {
      onDone$.next();
    }});
