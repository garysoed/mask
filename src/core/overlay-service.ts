import {source} from 'grapevine';
import {RenderSpec} from 'persona';
import {Observable, Subject} from 'rxjs';

export enum Anchor {
  START = 's',
  MIDDLE = 'm',
  END = 'e',
}

export interface AnchorSpec {
  readonly horizontal: Anchor;
  readonly vertical: Anchor;
}

export interface ShowEvent {
  readonly contentRenderSpec: RenderSpec;
  readonly contentAnchor: AnchorSpec;
  readonly target: Element;
  readonly targetAnchor: AnchorSpec;
}

export class OverlayService {
  readonly #onShow$ = new Subject<ShowEvent>();

  get onShow$(): Observable<ShowEvent> {
    return this.#onShow$;
  }

  show(event: ShowEvent): void {
    this.#onShow$.next(event);
  }
}

export const $overlayService = source(() => new OverlayService());
