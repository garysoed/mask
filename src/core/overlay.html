<style>
  #root {
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
  }

  #root.hidden {
    display: none;
  }

  #content {
    position: fixed;
  }
</style>
<div id="root">
  <div id="content" shadow="up">
    <!-- #content -->
  </div>
</div>
