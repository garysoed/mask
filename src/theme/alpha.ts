export enum Alpha {
  HIGH,
  MED_HIGH,
  MED_LOW,
  LOW,
  VERY_LOW,
}

export function getAlphaNumber(alpha: Alpha): number {
  switch (alpha) {
    case Alpha.HIGH:
      return 1;
    case Alpha.MED_HIGH:
      return 0.75;
    case Alpha.MED_LOW:
      return 0.65;
    case Alpha.LOW:
      return 0.45;
    case Alpha.VERY_LOW:
      return 0.35;
  }
}
