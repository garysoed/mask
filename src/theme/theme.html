<style>
  tr > td:first-child {
    padding-right: 2rem;
    text-align: end;
  }

  #col1 {
    width: 16rem;
  }

  #col2, #col3, #col4, #col5, #col6, #col7 {
    width: 12rem;
  }

  tbody tr {
    height: 8em;
  }

  thead th {
    text-align: center;
  }

  [mk-action-1].focus {
    --mkThemeBG: var(--mkThemeFocusBG1);
    --mkThemeFG: var(--mkThemeFocusFG1);
  }

  [mk-action-2].focus {
    --mkThemeBG: var(--mkThemeFocusBG2);
    --mkThemeFG: var(--mkThemeFocusFG2);
  }

  [mk-action-1].hover {
    --mkThemeBG: var(--mkThemeHoverBG1);
    --mkThemeFG: var(--mkThemeHoverFG1);
  }

  [mk-action-2].hover {
    --mkThemeBG: var(--mkThemeHoverBG2);
    --mkThemeFG: var(--mkThemeHoverFG2);
  }

  .label {
    font-size: 1rem;
  }

  .label.dark,
  #table.dark .label.light {
    display: none;
  }

  #table.dark .label.dark,
  .label.light {
    display: block;
  }
</style>

<table id="table" mk-theme-level="0">
  <colgroup>
    <col id="col1">
    <col id="col2">
    <col id="col3">
    <col id="col4">
    <col id="col5">
    <col id="col6">
    <col id="col7">
  </colgroup>
  <thead mk-body-2>
    <tr>
      <th></th>
      <th colspan="2">Main</th>
      <th colspan="2">Side</th>
      <th colspan="2">Highlight</th>
    </tr>
  </thead>
  <tbody>
    <!-- Passive -->
    <tr id="passiveRow">
      <td mk-body-2>passive</td>
      <td mk-passive-1 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-passive-2 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-passive-1 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-passive-2 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-passive-1 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-passive-2 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Action -->
    <tr id="actionRow">
      <td mk-body-2>action</td>
      <td mk-action-1 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Focus -->
    <tr id="focusRow">
      <td mk-body-2>action (focus)</td>
      <td mk-action-1 class="focus" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="focus" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="focus" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="focus" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="focus" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="focus" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Hover -->
    <tr id="hoverRow">
      <td mk-body-2>action (hover)</td>
      <td mk-action-1 class="hover" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="hover" mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="hover" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="hover" mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 class="hover" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 class="hover" mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>


    <!-- Disabled -->
    <tr id="disabledRow">
      <td mk-body-2>action (disabled)</td>
      <td mk-action-1 mk-disabled mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-disabled mk-theme-context="main">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-disabled mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-disabled mk-theme-context="side">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>

      <td mk-action-1 mk-disabled mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>A</h3>
        </div>
      </td>
      <td mk-action-2 mk-disabled mk-theme-context="highlight">
        <div flex-align="center" layout="column">
          <h3 flex mk-display>a</h3>
        </div>
      </td>
    </tr>
  </tbody>
</table>
