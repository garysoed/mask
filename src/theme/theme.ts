import { $asArray, $flat, $map, $pipe } from 'gs-tools/export/collect';
import { Color } from 'gs-tools/export/color';
import { cache } from 'gs-tools/export/data';

import generalCss from './general.css';
import { PRIMARY_DARK_HIGHLIGHT_SECTION_SPECS, PRIMARY_DARK_SECTION_SPECS, PRIMARY_LIGHT_HIGHLIGHT_SECTION_SPECS, PRIMARY_LIGHT_SECTION_SPECS, SECONDARY_DARK_HIGHLIGHT_SECTION_SPECS, SECONDARY_DARK_SECTION_SPECS, SECONDARY_LIGHT_HIGHLIGHT_SECTION_SPECS, SECONDARY_LIGHT_SECTION_SPECS, SectionSpec } from './section-spec';
import { ColorWithAlpha, createColor } from './shade';


export class Theme {
  constructor(
      private readonly document: Document,
      readonly baseColor: Color,
      readonly accentColor: Color,
  ) { }

  getStyleEl(): HTMLStyleElement {
    return this.createStyleEl();
  }

  setBaseColor(color: Color): Theme {
    return new Theme(this.document, color, this.accentColor);
  }

  setHighlightColor(color: Color): Theme {
    return new Theme(this.document, this.baseColor, color);
  }

  @cache()
  private createStyleEl(): HTMLStyleElement {
    const cssDeclarations = [
      this.generateSectionThemeCss(false),
      this.generateSectionThemeCss(true),
      this.generateHighlightThemeCss(),
    ];

    const styleEl = this.document.createElement('style');
    styleEl.innerHTML = `${cssDeclarations.join('\n\n')}\n${generalCss}`;
    return styleEl;
  }

  private generateHighlightThemeCss(): string {
    // Just use the light ones. The dark ones will be the same.
    const cssVariables = $pipe(
        [...PRIMARY_LIGHT_SECTION_SPECS, ...SECONDARY_LIGHT_SECTION_SPECS],
        $map(spec => {
          const cssNormalFgVar = getSectionVariableName(spec, true);
          const cssNormalBgVar = getSectionVariableName(spec, false);

          const highlightSpec = {...spec, isHighlight: true};
          const cssHighlightFgVar = getSectionVariableName(highlightSpec, true);
          const cssHighlightBgVar = getSectionVariableName(highlightSpec, false);

          return [
            `${cssNormalBgVar}: var(${cssHighlightBgVar});`,
            `${cssNormalFgVar}: var(${cssHighlightFgVar});`,
          ];
        }),
        $flat(),
        $asArray(),
    )
    .join('\n');

    return `[mk-theme-highlight] { ${cssVariables} }`;
  }

  private generateSectionThemeCss(isDark: boolean): string {
    const normalSectionSpecs = isDark ?
        [...PRIMARY_DARK_SECTION_SPECS, ...SECONDARY_DARK_SECTION_SPECS] :
        [...PRIMARY_LIGHT_SECTION_SPECS, ...SECONDARY_LIGHT_SECTION_SPECS];
    const highlightSectionSpecs = isDark ?
        [...PRIMARY_DARK_HIGHLIGHT_SECTION_SPECS, ...SECONDARY_DARK_HIGHLIGHT_SECTION_SPECS] :
        [...PRIMARY_LIGHT_HIGHLIGHT_SECTION_SPECS, ...SECONDARY_LIGHT_HIGHLIGHT_SECTION_SPECS];

    const cssVariables = $pipe(
        [...normalSectionSpecs, ...highlightSectionSpecs],
        $map(spec => {
          const base = spec.isBase ? this.baseColor : this.accentColor;
          const cssFgVar = getSectionVariableName(spec, true);
          const cssBgVar = getSectionVariableName(spec, false);
          const cssFgColor = generateCssColor(createColor(spec.fg, base));
          const cssBgColor = generateCssColor(createColor(spec.bg, base));
          return [
            `${cssFgVar}: ${cssFgColor};`,
            `${cssBgVar}: ${cssBgColor};`,
          ];
        }),
        $flat(),
        $asArray(),
    )
    .join('\n');

    const theme = isDark ? 'dark' : 'light';
    return `[mk-theme="${theme}"] { ${cssVariables} }`;
  }
}

function getSectionVariableName(sectionSpec: SectionSpec, isForeground: boolean): string {
  return [
    '--mkTheme',
    sectionSpec.section,
    isForeground ? 'FG' : 'BG',
    sectionSpec.isSecondary ? '2' : '1',
    sectionSpec.isHighlight ? 'H' : 'B',
  ].join('');
}

function generateCssColor({color, alpha}: ColorWithAlpha): string {
  return `rgba(${color.red},${color.green},${color.blue},${alpha})`;
}
