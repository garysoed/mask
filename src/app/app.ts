import {Vine, source} from 'grapevine';
import {PersonaBuilder} from 'persona';
import {BaseCtrlCtor, CustomElementCtrlCtor} from 'persona/export/internal';
import {BehaviorSubject} from 'rxjs';
import {map, pairwise, startWith} from 'rxjs/operators';

import {ThemeClassLoader} from '../theme/loader/theme-class-loader';
import {ThemeLoader} from '../theme/loader/theme-loader';
import {PALETTE} from '../theme/palette';
import {Theme} from '../theme/theme';


export const _p = new PersonaBuilder();

export const $themeLoader = source(
    'Theme',
    () => new BehaviorSubject<ThemeLoader>(
        new ThemeClassLoader(new Theme(PALETTE.ORANGE, PALETTE.GREEN))),
);
export const $window = source('window', () => window);

export function start(
    appName: string,
    rootCtrls: ReadonlyArray<CustomElementCtrlCtor|BaseCtrlCtor>,
    rootDoc: Document,
    themeLoader: ThemeLoader,
    customElementRegistry: CustomElementRegistry = window.customElements,
): {vine: Vine} {
  const vine = new Vine({appName});
  _p.build({rootCtrls, rootDoc, customElementRegistry, vine});
  const themeLoader$ = $themeLoader.get(vine);
  themeLoader$.next(themeLoader);
  themeLoader$
      .pipe(
          map(themeLoader => themeLoader.createElement(rootDoc)),
          startWith(null),
          pairwise(),
      )
      .subscribe(([oldEl, newEl]) => {
        if (oldEl) {
          oldEl.remove();
        }

        if (newEl) {
          rootDoc.head.appendChild(newEl);
        }
      });

  return {vine};
}
