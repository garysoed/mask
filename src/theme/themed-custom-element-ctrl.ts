import { CustomElementCtrl, PersonaContext } from 'persona';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';

import { $theme, _p } from '../app/app';


@_p.baseCustomElement({})
export abstract class ThemedCustomElementCtrl extends CustomElementCtrl {
  protected readonly theme$ = $theme.get(this.vine);

  constructor(context: PersonaContext) {
    super(context);

    this.addSetup(this.setupThemeUpdate());
  }

  private setupThemeUpdate(): Observable<unknown> {
    return this.theme$
        .pipe(
            tap(theme => {
              theme.injectCss(this.shadowRoot);
            }),
        );
  }
}
