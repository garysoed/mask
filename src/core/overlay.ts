import { cache } from 'gs-tools/export/data';
import { filterNonNull } from 'gs-tools/export/rxjs';
import { instanceofType } from 'gs-types';
import { classToggle, element, host, onDom, PersonaContext, resizeObservable, single, style } from 'persona';
import { combineLatest, merge, Observable, Subject } from 'rxjs';
import { debounceTime, delay, filter, map, mapTo, share, shareReplay, startWith, switchMap, tap, withLatestFrom } from 'rxjs/operators';
import { Logger } from 'santa';

import { _p } from '../app/app';
import { ThemedCustomElementCtrl } from '../theme/themed-custom-element-ctrl';

import { $overlayService, Anchor, NodeSpec, ShowEvent } from './overlay-service';
import template from './overlay.html';


const LOGGER = new Logger('mask.Overlay');

interface Position {
  readonly left: number;
  readonly top: number;
}


export const $overlay = {
  tag: 'mk-overlay',
  api: {},
};

export const $ = {
  host: host($overlay.api),
  content: element('content', instanceofType(HTMLDivElement), {
    content: single('#content'),
    styleTop: style('top'),
    styleLeft: style('left'),
  }),
  root: element('root', instanceofType(HTMLDivElement), {
    hidden: classToggle('hidden'),
    onClick: onDom('click'),
  }),
};

@_p.customElement({
  ...$overlay,
  template,
})
export class Overlay extends ThemedCustomElementCtrl {
  constructor(context: PersonaContext) {
    super(context);

    this.render($.root._.hidden, this.isRootHidden$);
    this.render($.content._.content, this.overlayContent$);
    this.render($.content._.styleLeft, this.contentLeft$);
    this.render($.content._.styleTop, this.contentTop$);
  }

  @cache()
  private get contentLeft$(): Observable<string> {
    return this.contentPosition$.pipe(
        map(position => {
          if (position === null) {
            return '';
          }
          return `${position.left}px`;
        }),
    );
  }

  @cache()
  private get contentPosition$(): Observable<Position|null> {
    return combineLatest([
      this.showStatus$,
      this.contentRect$,
    ])
    .pipe(
        map(([showStatus, contentRect]) => {
          if (!showStatus) {
            return null;
          }

          const targetRect = showStatus.target.node.getBoundingClientRect();
          const targetPosition = {
            left: computeAnchorLeft(targetRect, showStatus.target),
            top: computeAnchorTop(targetRect, showStatus.target),
          };
          const contentPosition = {
            left: computeAnchorLeft(contentRect, showStatus.content),
            top: computeAnchorTop(contentRect, showStatus.content),
          };

          return {
            left: contentRect.left - contentPosition.left + targetPosition.left,
            top: contentRect.top - contentPosition.top + targetPosition.top,
          };
        }),
        shareReplay({bufferSize: 1, refCount: true}),
    );
  }

  @cache()
  private get contentRect$(): Observable<DOMRect> {
    return this.declareInput($.content).pipe(
        switchMap(contentEl => {
          return resizeObservable(contentEl, {}).pipe(
              map(entries => {
                return entries[entries.length - 1]?.contentRect ?? null;
              }),
              filterNonNull(),
              startWith(contentEl.getBoundingClientRect()),
          );
        }),
        shareReplay({bufferSize: 1, refCount: false}),
    );
  }

  @cache()
  private get contentTop$(): Observable<string> {
    return this.contentPosition$.pipe(
        map(position => {
          if (position === null) {
            return '';
          }
          return `${position.top}px`;
        }),
    );
  }

  @cache()
  private get isRootHidden$(): Observable<boolean> {
    return this.showStatus$.pipe(map(showStatus => !showStatus));
  }

  @cache()
  private get overlayContent$(): Observable<Node|null> {
    return this.showStatus$.pipe(map(status => status?.content.node ?? null));
  }

  @cache()
  private get showStatus$(): Observable<ShowEvent|null> {
    const onShow$ = $overlayService.get(this.vine).pipe(
        switchMap(service => service.onShow$),
    );

    const onClick$ = this.declareInput($.root._.onClick).pipe(
        withLatestFrom(this.declareInput($.root)),
        filter(([event, rootEl]) => event.target === rootEl),
        mapTo(null),
    );

    return merge(onShow$, onClick$).pipe(
        startWith(null),
        shareReplay({bufferSize: 1, refCount: true}),
    );
  }
}

function computeAnchorLeft(elRect: DOMRect, nodeSpec: NodeSpec<any>): number {
  switch (nodeSpec.horizontal) {
    case Anchor.START:
      return elRect.left;
    case Anchor.MIDDLE:
      return elRect.left + elRect.width / 2;
    case Anchor.END:
      return elRect.right;
  }
}

function computeAnchorTop(elRect: DOMRect, nodeSpec: NodeSpec<any>): number {
  switch (nodeSpec.vertical) {
    case Anchor.START:
      return elRect.top;
    case Anchor.MIDDLE:
      return elRect.top + elRect.height / 2;
    case Anchor.END:
      return elRect.bottom;
  }
}
