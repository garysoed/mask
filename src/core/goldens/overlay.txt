<mk-overlay><style>
  #root {
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
  }

  #root.hidden {
    display: none;
  }

  #content {
    position: fixed;
  }
</style>
<div id="root" class="">
  <div id="content" shadow="up" style="left: 0px; top: 0px;">
    <!-- #content --><div id="testContent">test content</div>
  </div>
</div>
<link rel="stylesheet" href="base/test.css"></mk-overlay>